src/debug/debugService.ts

```diff
@@ -0,0 +1,251 @@
+                await this.waitForProxyReady(proxyResult.proxyProcess);
+                kubeChannel.showOutput("Finished to wait.");
+                
+                // Start debug session.
+                p.report({ message: `Starting ${this.debugProvider.getDebuggerType()} debug session...`});
+                kubeChannel.showOutput("Starting to start debug session...", "Start debug");
+                const sessionName = appName || `${Date.now()}`;
+                await this.startDebugging(cwd, sessionName, proxyResult.proxyDebugPort, proxyResult.proxyAppPort, async () => {
+                    proxyResult.proxyProcess.kill();
+                    if (appName) {
+                        await this.cleanupResource(`deployment/${appName}`);
+                    }
+                });
+            } catch (error) {
+                vscode.window.showErrorMessage(error);
+                kubeChannel.showOutput(`Debug(Launch) on kubernetes failed. See the errors: ${error}.`);
+                if (appName) {
+                    await this.cleanupResource(`deployment/${appName}`);
+                }
+            }
+            return null;
+        });
+    }
+
+    /**
+     * In attach mode, the user should choose the running pod first, then the debugger will attach to it.
+     * 
+     * @param workspaceFolder the active workspace folder.
+     * @param pod the debug pod name.
+     */
+    public async attachDebug(workspaceFolder: vscode.WorkspaceFolder, pod?: string): Promise<void> {
+        // TODO
+        return;
+    }
+
+    private async createPortForward(podName, debugPort, appPort): Promise<any> {
+        const portMapping = [];
+        const portfinder = require('portfinder');
+        // Find a free local port for forwarding data to remote app port.
+        let proxyAppPort = 0;
+        if (appPort) {
+            proxyAppPort = await portfinder.getPortPromise({
+                port: appPort
+            });
+            portMapping.push(proxyAppPort + ":" + appPort);
+        }
+        // Find a free local port for forwarding data to remote debug port.
+        const proxyDebugPort = await portfinder.getPortPromise({
+            port: Math.max(10000, Number(proxyAppPort) + 1)
+        });
+        portMapping.push(proxyDebugPort + ":" + debugPort);
+
+        let bin = vscode.workspace.getConfiguration('vs-kubernetes')['vs-kubernetes.kubectl-path'];
+        if (!bin) {
+            bin = 'kubectl';
+        }
+        
+        return {
+            proxyProcess: require('child_process').spawn(bin, ["port-forward", podName, ...portMapping]),
+            proxyDebugPort,
+            proxyAppPort
+        };
+    }
+
+    private async waitForProxyReady(proxyProcess: ChildProcess): Promise<void> {
+        return new Promise<void>((resolve, reject) => {
+            const forwardingRegExp = /Forwarding\s+from\s+127\.0\.0\.1:/;

```

Comment:
In regular expression, the space should be `\s`, and adding an additional `+` does no harm. Right?

Reply:
Well, it means you need a regex instead of a string contains check, and regexes are more error-prone to maintain.  But it's no big deal for this simple case!